hwilib/devices/tapsigner.py

```diff
@@ -0,0 +1,474 @@
+"""
+Tapsigner
+********
+"""
+import os
+import pyaes
+import struct
+import getpass
+import datetime
+from base64 import b64encode
+from typing import List, Dict, Any, Union, Tuple, Sequence
+from cktap.transport import find_cards, CKTapUnixTransport, CKTapCard, CardRuntimeError  # type: ignore
+from cktap.utils import ser_compact_size, str2path, path2str, pick_nonce  # type: ignore
+from ..common import Chain, AddressType, hash256, hash160
+from ..key import ExtendedKey, get_bip44_purpose, get_bip44_chain, HARDENED_FLAG
+from ..psbt import PSBT
+from ..descriptor import MultisigDescriptor
+from ..errors import UnavailableActionError, BadArgumentError
+from ..hwwclient import HardwareWalletClient
+from .._serialize import ser_sig_der, ser_string
+from .._script import (
+    is_p2pk,
+    is_p2pkh,
+    is_p2sh,
+    is_p2wpkh,
+    is_p2wsh,
+    is_witness,
+)
+
+
+def is_cvc_length_range(cvc: str) -> bool:
+    """CVC has to be at least 6 characters long and at most 32 characters long"""
+    return 6 <= len(cvc) <= 32
+
+
+def find_card_by_path(path: str) -> CKTapCard:
+    # path is actually an ident
+    for card in find_cards():
+        if path in card.card_ident:
+            return card
+    raise RuntimeError("Card not found! Is it on the reader?")
+
+
+def is_hardened(component: int) -> bool:
+    return component >= HARDENED_FLAG
+
+
+def check_bip32_path(path: Sequence[int]) -> None:
+    """Do not allow hardened components after non-hardened"""
+    found_non_hardened = False
+    for component in path:
+        curr_hardened = is_hardened(component)
+        # have we found non hardened component?
+        curr_non_hardened = not curr_hardened
+        if curr_hardened is True and found_non_hardened is True:
+            raise ValueError(f"Hardened path component after non-hardened {path}")
+        found_non_hardened = curr_non_hardened
+
+
+def split_bip32_path(path: Sequence[int]) -> Tuple[List[int], List[int]]:
+    """Split bip32 path to hardened part and non-hardened part"""
+    hardened = []
+    non_hardened = []
+    for component in path:
+        if is_hardened(component):
+            hardened.append(component)
+        else:
+            non_hardened.append(component)
+    return hardened, non_hardened
+
+
+class TapsignerClient(HardwareWalletClient):
+    # In path variable we store card ident
+    # In password variable we store card CVC
+    def __init__(self, path: str, password: str = "", expert: bool = False, chain: Chain = Chain.MAIN) -> None:
+        super(TapsignerClient, self).__init__(path, password, expert, chain)
+        self.device = find_card_by_path(path=path)
+
+    def interactive_get_cvc(self) -> None:
+        if not self.password:
+            cvc = getpass.getpass("Please provide CVC:\n")
+            if cvc:
+                self.password = cvc
```

Comment:
We prefer to validate that the passphrase was provided in `__init__` rather than during the individual functions and interactively retrieving it. Having this optional interactivity breaks our model of non-interactivity for most functions.

Reply:
check for password in init - raise if not provided; remove interactive_get_cvc